File: backend_python/src/controllers/archivos_controller.py
import boto3
import base64
import re
import time
import os
from fastapi import HTTPException
from dotenv import load_dotenv

load_dotenv()

# 1. Configuración dinámica
s3_client = boto3.client(
    's3',
    region_name=os.getenv("AWS_REGION"),
    endpoint_url=os.getenv("AWS_ENDPOINT_URL"),
    aws_access_key_id=os.getenv("AWS_ACCESS_KEY_ID"),
    aws_secret_access_key=os.getenv("AWS_SECRET_ACCESS_KEY")
)

def subir_foto_perfil(base64_image: str):
    if not base64_image:
        raise HTTPException(status_code=400, detail="No se proporcionó ninguna imagen")

    # 2. Limpieza robusta del Base64
    match = re.match(r"^data:(image/[A-Za-z-+\/]+);base64,(.+)$", base64_image)
    
    if not match:
        raise HTTPException(status_code=400, detail="Formato de imagen inválido")

    content_type = match.group(1)
    base64_data = match.group(2)
    image_buffer = base64.b64decode(base64_data)

    # 3. Crear nombre único
    extension = content_type.split('/')[1] if '/' in content_type else 'jpg'
    file_name = f"perfil-{int(time.time() * 1000)}.{extension}"
    bucket_name = os.getenv("AWS_BUCKET_NAME")

    try:
        # 4. Subir a MinIO / S3
        s3_client.put_object(
            Bucket=bucket_name,
            Key=file_name,
            Body=image_buffer,
            ContentType=content_type
        )

        # 5. Construir la URL
        image_url = f"{os.getenv('AWS_ENDPOINT_URL')}/{bucket_name}/{file_name}"
        
        return {"mensaje": "Imagen subida con éxito", "url": image_url}
    
    except Exception as e:
        print(f"Error S3: {e}")
        raise HTTPException(status_code=500, detail="Error interno al guardar la imagen")

File: backend_node/src/controllers/archivos.controller.js
const { S3Client, PutObjectCommand } = require("@aws-sdk/client-s3");

// 1. Configuración dinámica usando tu .env
const s3Client = new S3Client({
  region: process.env.AWS_REGION,
  endpoint: process.env.AWS_ENDPOINT_URL,
  credentials: {
    accessKeyId: process.env.AWS_ACCESS_KEY_ID,
    secretAccessKey: process.env.AWS_SECRET_ACCESS_KEY,
  },
  forcePathStyle: true,
});

const subirFotoPerfil = async (req, res) => {
  try {
    const { base64Image } = req.body;

    if (!base64Image) {
      return res
        .status(400)
        .json({ mensaje: "No se proporcionó ninguna imagen" });
    }

    // 2. Limpieza robusta del Base64 (Soporta PNG, JPEG, WEBP, etc.)
    const matches = base64Image.match(/^data:([A-Za-z-+\/]+);base64,(.+)$/);

    if (!matches || matches.length !== 3) {
      return res.status(400).json({ mensaje: "Formato de imagen inválido" });
    }

    const contentType = matches[1];
    const base64Data = matches[2];
    const buffer = Buffer.from(base64Data, "base64");

    // 3. Crear nombre único y leer el bucket del .env
    const extension = contentType.split("/")[1] || "jpg";
    const fileName = `perfil-${Date.now()}.${extension}`;
    const bucketName = process.env.AWS_BUCKET_NAME;

    // 4. Subir a MinIO / S3
    const command = new PutObjectCommand({
      Bucket: bucketName,
      Key: fileName,
      Body: buffer,
      ContentType: contentType,
    });

    await s3Client.send(command);

    // 5. Construir la URL pública dinámicamente
    const imageUrl = `${process.env.AWS_ENDPOINT_URL}/${bucketName}/${fileName}`;

    res.status(200).json({
      mensaje: "Imagen subida con éxito",
      url: imageUrl,
    });
  } catch (error) {
    console.error("Error al subir a S3/MinIO:", error);
    res.status(500).json({ mensaje: "Error interno al guardar la imagen" });
  }
};

module.exports = { subirFotoPerfil };
